fabric-api-base/src/main/java/net/fabricmc/fabric/api/event/EventFactory.java

```diff
@@ -47,18 +47,70 @@ public static void invalidate() {
 
 	/**
 	 * Create an "array-backed" Event instance.
+	 * The factory will be used for any number of listeners.
+	 * Consider using {@link #createArrayBackedCustomEmptyInvoker(Class, Object, Function)}  the more optimized overload} if you need a slight performance increase
+	 * when there are 0 or 1 listeners.
 	 *
 	 * @param type           The listener class type.
 	 * @param invokerFactory The invoker factory, combining multiple listeners into one instance.
 	 * @param <T>            The listener type.
 	 * @return The Event instance.
 	 */
+	public static <T> Event<T> createArrayBackedUsingTheFactoryEveryTime(Class<? super T> type, Function<T[], T> invokerFactory) {
+		return EventFactoryImpl.createArrayBackedUsingTheFactoryEveryTime(type, invokerFactory);
+	}
+
+	/**
+	 * Create an "array-backed" Event instance, with a custom empty invoker.
+	 * The empty invoker will be used when there are no listeners,
+	 * and when there is only one listener, it will be used directly.
+	 * The factory will only be used when there are at least two listeners.
+	 * Consider using {@link #createArrayBackedUsingTheFactoryEveryTime} if this optimization is unsuitable or unneeded.
+	 *
+	 * @param type           The listener class type.
+	 * @param emptyInvoker   The custom empty invoker.
+	 * @param invokerFactory The invoker factory, combining multiple listeners into one instance.
+	 * @param <T>            The listener type.
+	 * @return The Event instance.
+	 */
+	public static <T> Event<T> createArrayBackedCustomEmptyInvoker(Class<? super T> type, T emptyInvoker, Function<T[], T> invokerFactory) {
```

Comment:
Why? Why can't the invoker factory handle the case when the input array is empty?

Reply:
Apparently it's slightly faster for events, which could be useful for events that are not used but are in a hot path. Arguably this is useless as soon as there is more than 1 listener.